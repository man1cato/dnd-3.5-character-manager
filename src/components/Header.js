import React from 'react';
import {connect} from 'react-redux';
import {startLogout} from '../actions/auth';

export const Header = (props) => (
  <header className="header">
    <div className="container container--header header__content">
      <h2 className="header__title">{props.pageTitle}</h2>
        <button className="button button--link" onClick={props.startLogout}>Logout</button>
    </div>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
