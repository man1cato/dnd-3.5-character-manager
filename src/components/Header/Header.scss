@import '../../styles/base/settings';


header {
    background: $dark-grey;
    position: fixed;
    top: 0;
    height: 8vh;
    min-height: 50px;
    width: -webkit-fill-available;
    padding: $s-size 0;
    box-shadow: 0 0 5px $light-grey;
}

.header__content {
    display: flex;
    align-items: center;
    justify-content: space-between;
    & * {
        color: white;
        background: none;
        border: none;
    }
}

.header__title {
    margin: 0;
    @media(max-width: 576px) {
      font-size: $font-size-l;
    }
}

.header__nav {
    display: flex;
    align-items: center;
    justify-content: flex-end;
    height: 100%;
    max-width: 60%;
}

.header__button {
    display: flex;
    align-items: center;
    padding: 0;
    margin-left: $s-size;
}