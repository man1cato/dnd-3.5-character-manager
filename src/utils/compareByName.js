export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}