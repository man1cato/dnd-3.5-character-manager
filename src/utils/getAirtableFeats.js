import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset
		let feats = []

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			const featGroup = res.data.records.map((feat) =>  ({
				id: feat.id,
				name: feat.fields.Name,
				type: feat.fields.Type.join(", ") || null,
				description: feat.fields.Description || null
			}));
			
			feats = feats.concat(featGroup)
		} while (!!offset)
		
		return _.orderBy(feats, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}