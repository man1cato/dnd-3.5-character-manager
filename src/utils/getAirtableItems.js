import dotenv from 'dotenv'
dotenv.config()

import axios from 'axios'
import _ from 'lodash'

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'

export default async () => {
	try {
		let offset;
		let items = [];
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			const itemGroup = res.data.records.map((item) =>  ({
				id: item.id,
				name: item.fields.Name,
				value: item.fields["Value (gp)"] || '',
				weight: item.fields["Weight (lbs)"] || '',
				category: item.fields.Category || null,
				weaponType: item.fields["Weapon Type"] || null,
				encumbrance: item.fields.Encumbrance || null,
				damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
				damageS: item.fields["Damage (S)"] || null,
				damageM: item.fields["Damage (M)"] || null,
				critical: item.fields.Critical || null,
				range: item.fields.Range || null,
				armorBonus: item.fields["Armor Bonus"] || null,
				maxDexBonus: item.fields["Max DEX Bonus"] || null
			}))
			
			items = items.concat(itemGroup)
		} while (!!offset)
		
		return _.orderBy(items, ['name'], ['asc'])
	} catch (e) {
		console.error(e)
	}
}