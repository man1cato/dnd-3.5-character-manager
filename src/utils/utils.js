import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])