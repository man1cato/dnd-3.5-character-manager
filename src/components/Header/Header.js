import React from 'react'
import { connect } from 'react-redux'
import { startLogout } from '../../store/actions/auth'
import { NavLink } from 'react-router-dom'

import './Header.scss'


export const Header = ({pageTitle, startLogout, hasProfiles}) => (
	<header>
		<div className="container header__content">
			<h2 className="header__title">{pageTitle}</h2>
			<div className="header__nav">
				<NavLink to="/create" className="header__button">
					<ion-icon name="add-circle-outline" size="large" />
				</NavLink>
				<NavLink to="/select" className="header__button" hidden={!hasProfiles}>
					<ion-icon name="swap" size="large" />
				</NavLink>
				<button className="header__button" onClick={startLogout}>
					<ion-icon name="log-out" size="large"/>
				</button>
			</div>
		</div>
	</header>
)

const mapStateToProps = (state) => ({
	hasProfiles: !!state.profiles && state.profiles.length > 1
})

const mapDispatchToProps = (dispatch) => ({
	startLogout: () => dispatch(startLogout())
})

export default connect(mapStateToProps, mapDispatchToProps)(Header)
