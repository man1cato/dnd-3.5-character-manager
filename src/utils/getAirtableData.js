import axios from 'axios'
import _ from 'lodash'
import { convertTextToArray } from './utils' 
import dotenv from 'dotenv'
dotenv.config({ path: './config/dev.env' })

const apiKey = process.env.AIRTABLE_API_KEY
const baseUrl = 'https://api.airtable.com/v0/appK7TZeddGqjGUDL'


export const getFeats = async () => {
	try {
		let offset
		let feats = {}

		do {
			const res = await axios.get(`${baseUrl}/Feats?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((feat) => {
				feats[feat.id] = {
					name: feat.fields.Name,
					types: feat.fields.Types || null,
					description: feat.fields.Description || null,
					prerequisites: {
						level: feat.fields['Prereq. Level'] || null,
						class: feat.fields['Prereq. Class'] || null,
						feats: feat.fields['Prereq. Feats'] || null
					} || null
				}
			})			
		} while (!!offset)
		
		return feats
	} catch (e) {
		console.error(e)
	}
}


export const getItems = async () => {
	try {
		let offset
		let items = {}
		do {
			const res = await axios.get(`${baseUrl}/Items?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
			offset = res.data.offset
			res.data.records.forEach((item) => {
				items[item.id] = {
					name: item.fields.Name,
					value: item.fields["Value (gp)"] || '',
					weight: item.fields["Weight (lbs)"] || '',
					category: item.fields.Category || null,
					weaponType: item.fields["Weapon Type"] || null,
					encumbrance: item.fields.Encumbrance || null,
					damageType: item.fields["Damage Type"] ? item.fields["Damage Type"].join(" / ") : [],
					damageS: item.fields["Damage (S)"] || null,
					damageM: item.fields["Damage (M)"] || null,
					critical: item.fields.Critical || null,
					range: item.fields.Range || null,
					armorBonus: item.fields["Armor Bonus"] || null,
					maxDexBonus: item.fields["Max DEX Bonus"] || null
				}
			})
		} while (!!offset)
		
		return items
	} catch (e) {
		console.error(e)
	}
}


const getLevels = async (jobClassId) => {
	try {
		const levelRes = await axios.get(`${baseUrl}/Levels?filterByFormula={Class ID}="${jobClassId}"&api_key=${apiKey}`)
		
		const levelRecords = levelRes.data.records
		let levels = {}
		for (let i = 0; i < levelRecords.length; i++) {
			const levelRecord = levelRecords[i].fields 
			levels[levelRecord.Level] = {
				saves: {
					fortitude: levelRecord['Fortitude'],
					reflex: levelRecord['Reflex'],
					will: levelRecord['Will']
				},
				baseAttackBonuses: _.compact([ levelRecord['BAB 1'], levelRecord['BAB 2'], levelRecord['BAB 3'], levelRecord['BAB 4'] ]),
				xp: levelRecord['XP Required'],
				specialAbilities: !!levelRecord['Special Abilities'] ? levelRecord['Special Abilities'] : null,
				spellsPerDay:  _.compact([ 
					levelRecord['SPD 0'], levelRecord['SPD 1'], levelRecord['SPD 2'], levelRecord['SPD 3'], levelRecord['SPD 4'], 
					levelRecord['SPD 5'], levelRecord['SPD 6'], levelRecord['SPD 7'], levelRecord['SPD 8'], levelRecord['SPD 9'] 
				])
			}
		}
		return levels
	} catch (e) {
		console.error(e)
	}
}

export const getJobClasses = async () => {
	try {
		let offset
		let jobClasses = {}
	
		do {
			const classRes = await axios.get(`${baseUrl}/Classes?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = classRes.data.offset

			const jobClassGroup = classRes.data.records
			
			for (let i = 0; i < jobClassGroup.length; i++) {
				const jobClass = jobClassGroup[i] 
				jobClasses[jobClass.id] = {
					name: jobClass.fields.Name,
					hitDie: jobClass.fields['Hit Die'],
					proficiencies: jobClass.fields['Proficiencies'] || [],
					skills: jobClass.fields.Skills || [],
					levels: await getLevels(jobClass.id)
				}
			}
		} while (!!offset)
		
		return jobClasses
	} catch (e) {
		console.error(e)
	}
}


export const getRaces = async () => {
	try {
		let offset
		let races = {}
		do {
			const res = await axios.get(`${baseUrl}/Races?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
			offset = res.data.offset
			res.data.records.forEach((race) =>  {
				races[race.id] = {
					name: race.fields.Name,
					size: race.fields.Size,
					speed: race.fields['Speed (ft/rnd)'],
					racialBonuses: race.fields['Racial Bonuses'] ? convertTextToArray(race.fields['Racial Bonuses']) : null,
					defaultLanguages: race.fields['Languages (Default)'],
					bonusLanguages: race.fields['Languages (Bonus)'],
					abilityMods: {
						str: race.fields['STR Mod'],
						dex: race.fields['DEX Mod'],
						con: race.fields['CON Mod'],
						int: race.fields['INT Mod'],
						wis: race.fields['WIS Mod'],
						cha: race.fields['CHA Mod']
					},
					specialAbilities: race.fields['Special Abilities'] || [],
					favoredClass:  race.fields['Favored Class'] ? race.fields['Favored Class'][0] : null,
					iconUrl: race.fields['Icon URL']
				}
			})
			
		} while (!!offset)
		
		return races
	} catch (e) {
		console.error(e)
	}
}


export const getSkills = async () => {
   let offset
   let skills = {}
   do {
		const res = await axios.get(`${baseUrl}/Skills?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '') )
		offset = res.data.offset
		res.data.records.forEach((skill) => {
			skills[skill.id] = {
				name: skill.fields.Name,
				keyAbility: skill.fields["Key Ability"] || null,
				description: skill.fields.Description || "",
				notes: skill.fields.Notes || ""
			}
		})
   } while (!!offset)
   
   return skills
}


export const getSpecialAbilities = async () => {
   let offset
   let specialAbilities = {}
   do {
		const res = await axios.get(`${baseUrl}/Special Abilities?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || ''))
		offset = res.data.offset
		res.data.records.forEach((specialAbility) => {
			specialAbilities[specialAbility.id] = {
				name: specialAbility.fields.Name,
				type: specialAbility.fields.Type ? specialAbility.fields.Type.join(", ") : null,
				description: specialAbility.fields.Description				
			}
		})
   } while (!!offset)
   
   return specialAbilities
}


export const getSpells =  async () => {
   let offset
   let spells = {}
   do {
		const res = await axios.get(`${baseUrl}/Spells?api_key=${apiKey}` + (!!offset && `&offset=${offset}` || '')) 
		offset = res.data.offset
		res.data.records.forEach((spell) => {
			spells[spell.id] = {
				name: spell.fields.Name,
				description: spell.fields.Description || "",
				school: spell.fields.School || null,
				components: spell.fields.Components.join(", ") || null,
				castingTime: spell.fields["Casting Time"] || null,
				range: spell.fields.Range || null,
				effectArea: spell.fields["Target, Effect, Area"] || null,
				duration: spell.fields.Duration || null,
				savingThrow: spell.fields["Saving Throw"] || null,
				spellResistance: spell.fields["Spell Resistance"] || null
			}
		})
   } while (!!offset)

   return spells
}