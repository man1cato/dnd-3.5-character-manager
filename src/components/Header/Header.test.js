import React from 'react'
import { fireEvent } from '@testing-library/react'

import { renderWithRedux } from '../../tests/utils'
import { Header } from './Header'


const startLogout = jest.fn()
let props
beforeAll(() => {
    props = {
        pageTitle: 'Profile',
        startLogout
    }
})


test('should render Header with swap button hidden', () => {
    const { container, getByTestId } = renderWithRedux(<Header {...props}/>)
    expect(container.firstChild).toMatchSnapshot()
    expect(getByTestId('swapButton')).toHaveProperty('hidden', true)
})

test('should call startLogout on button click', () => {
    const { getByTestId } = renderWithRedux(<Header {...props} />)
    const logoutButton = getByTestId('logoutButton')
    fireEvent.click(logoutButton)
    expect(startLogout).toHaveBeenCalled()
})

test('should show select character icon when profiles in store', () => {
    const { getByTestId } = renderWithRedux(<Header {...props} hasProfiles />)
    expect(getByTestId('swapButton')).toHaveProperty('hidden', false)
})