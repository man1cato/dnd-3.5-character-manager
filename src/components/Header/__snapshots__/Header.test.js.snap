// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Header with swap button hidden 1`] = `
<header>
  <div
    class="container header__content"
  >
    <h2
      class="header__title"
    >
      Profile
    </h2>
    <div
      class="header__nav"
    >
      <a
        class="header__button"
        href="/create"
      >
        <ion-icon
          name="add-circle-outline"
          size="large"
        />
      </a>
      <a
        class="header__button"
        data-testid="swapButton"
        hidden=""
        href="/select"
      >
        <ion-icon
          name="swap"
          size="large"
        />
      </a>
      <button
        class="header__button"
        data-testid="logoutButton"
      >
        <ion-icon
          name="log-out"
          size="large"
        />
      </button>
    </div>
  </div>
</header>
`;
